Source/AMRInterpolator/SphericalExtraction.hpp

```diff
@@ -0,0 +1,139 @@
+        std::pair<std::vector<double>, std::vector<double>> &out_integrals,
+        const IntegrationMethod &a_method_theta = IntegrationMethod::simpson,
+        const IntegrationMethod &a_method_phi = IntegrationMethod::trapezium)
+    {
+        auto integrand_re = [center = m_center, &geom = m_geom, es, el, em,
+                             &a_function](std::vector<double> &a_data_here,
+                                          double r, double theta, double phi) {
+            // note that spin_Y_lm requires the coordinates with the center
+            // at the origin
+            double x = geom.get_grid_coord(0, r, theta, phi) - center[0];
+            double y = geom.get_grid_coord(1, r, theta, phi) - center[1];
+            double z = geom.get_grid_coord(2, r, theta, phi) - center[2];
+            SphericalHarmonics::Y_lm_t<double> Y_lm =
+                SphericalHarmonics::spin_Y_lm(x, y, z, es, el, em);
+            auto function_here = a_function(a_data_here, r, theta, phi);
+            return (function_here.first * Y_lm.Real +
+                    function_here.second * Y_lm.Im) /
+                   (r * r);
+        };
+        add_integrand(integrand_re, out_integrals.first, a_method_theta,
+                      a_method_phi);
+
+        auto integrand_im = [center = m_center, &geom = m_geom, es, el, em,
+                             &a_function](std::vector<double> &a_data_here,
+                                          double r, double theta, double phi) {
+            // note that spin_Y_lm requires the coordinates with the center
+            // at the origin
+            double x = geom.get_grid_coord(0, r, theta, phi) - center[0];
+            double y = geom.get_grid_coord(1, r, theta, phi) - center[1];
+            double z = geom.get_grid_coord(2, r, theta, phi) - center[2];
+            SphericalHarmonics::Y_lm_t<double> Y_lm =
+                SphericalHarmonics::spin_Y_lm(x, y, z, es, el, em);
+            auto function_here = a_function(a_data_here, r, theta, phi);
+            return (function_here.second * Y_lm.Real -
+                    function_here.first * Y_lm.Im) /
+                   (r * r);
+        };
+        add_integrand(integrand_im, out_integrals.second, a_method_theta,
+                      a_method_phi);
+    }
+
+    //! If you only want to extract one mode, you can use this function which
+    //! calls add_mode_integrand, then integrate and returns the integrals
+    std::pair<std::vector<double>, std::vector<double>> integrate_mode(
+        int es, int el, int em, complex_function_t a_function,
+        const IntegrationMethod &a_method_theta = IntegrationMethod::simpson,
+        const IntegrationMethod &a_method_phi = IntegrationMethod::trapezium)
+    {
+        std::pair<std::vector<double>, std::vector<double>> integrals;
+        add_mode_integrand(es, el, em, a_function, integrals, a_method_theta,
+                           a_method_phi);
+        integrate();
```

Comment:
I may be missing something, but if I want to integrate several modes and, instead of adding them with `add_mode_integrand` and then integrate, I just call this `integrate_mode` several times, this line here with `integrate` will cause it to integrate all the previous additions again, right? If I'm right, an easy solution is to use the `integrate` overload from `SurfaceExtraction` that receives the integrand directly (as that one clears previous integrands)

Reply:
Ah yes, I hadn't spotted that. Unfortunately your solution involves more code duplication so instead I will `clear` the relevant `std::vector`s at the start of this function.